operators/src/util/number_statistics.rs

```diff
@@ -0,0 +1,150 @@
+use num_traits::AsPrimitive;
+use serde::{Deserialize, Serialize};
+
+/// This struct provides some basic number statistics.
+///
+/// All operations run in constant time.
+#[derive(Debug, Clone, Copy, PartialEq, Serialize, Deserialize)]
+pub struct NumberStatistics {
+    min_value: f64,
+    max_value: f64,
+    value_count: usize,
+    value_nan_count: usize,
+    mean_value: f64,
+    m2: f64,
+}
+
+impl Default for NumberStatistics {
+    fn default() -> Self {
+        Self {
+            min_value: f64::MAX,
+            max_value: f64::MIN,
+            value_count: 0,
+            value_nan_count: 0,
+            mean_value: 0.0,
+            m2: 0.0,
+        }
+    }
+}
+
+impl NumberStatistics {
+    #[inline]
+    pub fn add<V>(&mut self, value: V)
+    where
+        V: AsPrimitive<f64>,
+    {
+        let value = value.as_();
+
+        if value.is_nan() {
+            self.value_nan_count += 1;
+            return;
+        }
+
+        self.min_value = f64::min(self.min_value, value);
+        self.max_value = f64::max(self.max_value, value);
+
+        // Welford's algorithm
+        self.value_count += 1;
+        let delta = value - self.mean_value;
+        self.mean_value += delta / (self.value_count as f64);
+        let delta2 = value - self.mean_value;
+        self.m2 += delta * delta2;
+    }
+
+    #[inline]
+    pub fn add_no_data(&mut self) {
+        self.value_nan_count += 1;
+    }
+
+    pub fn count(&self) -> usize {
+        self.value_count
+    }
+
+    pub fn nan_count(&self) -> usize {
+        self.value_nan_count
+    }
+
+    pub fn min(&self) -> f64 {
+        if self.value_count > 0 {
+            self.min_value
+        } else {
+            f64::NAN
+        }
+    }
+
+    pub fn max(&self) -> f64 {
+        if self.value_count > 0 {
+            self.max_value
+        } else {
+            f64::NAN
+        }
+    }
+
+    pub fn mean(&self) -> f64 {
```

Comment:
why return NAN and not Option<f64> ?

Reply:
It only occurs on very little data, I'm not sure, do you think it is worthwhile distinguishing?

Reply:
I think for rasters it will occure on a lot of data. I think for the API it's fine to ouput NAN. Not sure what's better in backend code. We can leave it if you think it's the correct semantics

Reply:
cf. our conversation, we will leave it for now like this.